watcher/watcher.go

```diff
@@ -16,6 +16,97 @@ import (
 	"time"
 )
 
+type pollWatcher struct {
+	paths map[string]bool
+}
+
+func (w *pollWatcher) isWatching(path string) bool {
+	a, b := w.paths[path]
+	return a && b
+}
+
+func (w *pollWatcher) Add(path string) error {
+	if w.paths == nil {
+		w.paths = map[string]bool{}
+	}
+	w.paths[path] = true
+	return nil
+}
+
+func (p *Project) watchByPolling() {
+	var wr sync.WaitGroup
+	var watcher = new(pollWatcher)
+	channel, exit := make(chan bool, 1), make(chan bool, 1)
+	p.path = p.Path
+	defer func() {
+		wg.Done()
+	}()
+	p.cmd(exit)
+	if err := p.walks(watcher); err != nil {
+		log.Fatalln(p.pname(p.Name, 2), ":", p.Red.Bold(err.Error()))
+		return
+	}
+	go p.routines(channel, &wr)
+	p.LastChangedOn = time.Now().Truncate(time.Second)
+	// waiting for an event
+
+	var walk = func(changed string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		} else if !watcher.isWatching(changed) {
+			return nil
+		} else if !info.ModTime().Truncate(time.Second).After(p.LastChangedOn) {
+			return nil
+		}
+
+		var ext string
+		if index := strings.Index(filepath.Ext(changed), "_"); index == -1 {
+			ext = filepath.Ext(changed)
+		} else {
+			ext = filepath.Ext(changed)[0:index]
+		}
+		i := strings.Index(changed, filepath.Ext(changed))
+		file := changed[:i] + ext
+		path := filepath.Dir(changed[:i])
+		if changed[:i] != "" && inArray(ext, p.Watcher.Exts) {
+			p.LastChangedOn = time.Now().Truncate(time.Second)
+			msg := fmt.Sprintln(p.pname(p.Name, 4), ":", p.Magenta.Bold(strings.ToUpper(ext[1:]) + " changed"), p.Magenta.Bold(file))
+			out := BufferOut{Time: time.Now(), Text: strings.ToUpper(ext[1:]) + " changed " + file}
+			p.print("log", out, msg, "")
+			// stop and run again
+			if p.Run {
+				close(channel)
+				channel = make(chan bool)
+			}
+			// handle multiple errors, need a better way
+			p.fmt(file)
+			p.test(path)
+			p.generate(path)
+			go p.routines(channel, &wr)
+		}
+
+		return nil
+	}
+	for {
+		for _, dir := range p.Watcher.Paths {
+			base := filepath.Join(p.base, dir)
+			if _, err := os.Stat(base); err == nil {
+				if err := filepath.Walk(base, walk); err != nil {
+					log.Println(p.Red.Bold(err.Error()))
+				}
+			} else {
+				log.Println(p.Red.Bold(base + " path doesn't exist"))
+			}
+
+			select {
+			case <-exit:
+				return
+			case <-time.After(p.parent.Config.PollingInterval / time.Duration(len(p.Watcher.Paths))):
```

Comment:
The problem is that people might watch files not in different directories, e.g. ['./src', '../../foo'], so either we have to start walking from the lowest common denominator, or recursively for each path (although this might contain duplicates).

I'm not sure how nodemon/chokidar solve this at the moment. Any ideas?